bot/services/pin_service.py

```diff
@@ -0,0 +1,104 @@
+from bot.sock_bot import SockBot
+from bot.utils.helpers import fetch_optional_message
+
+PIN_REACTION = '📌'
+MIN_PIN_REACTIONS = 5
+MAX_CONTENT_CHARS = 250
+MAX_PINS_PER_CHANNEL = 50
+
+
+class PinService(BaseService):
+
+    def __init__(self, *, bot: SockBot):
+        super().__init__(bot)
+        self.pin_repo = PinRepository()
+        self.class_repo = ClassRepository()
+
+    @BaseService.Listener(Events.on_pin_request)
+    async def on_pin_request(self, ctx: Context, message: discord.Message):
+        # format our message content
+        if len(message.content) > MAX_CONTENT_CHARS:
+            content = message.content[:MAX_CONTENT_CHARS] + '...'
+        else:
+            content = message.content
+        # prepare our embed
+        embed = discord.Embed(title='📌 Pin Request', color=Colors.Purple)
+        embed.description = f'{ctx.author.mention} wants to pin a message.\n' \
+                            f'Click the {PIN_REACTION} reaction below to pin this message.'
+        embed.add_field(name='Content', value=content, inline=False)
+        embed.add_field(name='Author', value=message.author.mention)
+        embed.add_field(name='Message Link', value=f'[Link]({message.jump_url})')
+        if len(message.attachments):
+            embed.add_field(name='Files Attached', value=len(message.attachments))
+        sockbot_message = await message.channel.send(embed=embed)
+        await sockbot_message.add_reaction(PIN_REACTION)
+        # create our class pin object and push it to the db
+        class_pin = ClassPin(sockbot_message.id, message.id, message.channel.id, message.author.id, ctx.author.id)
+        await self.pin_repo.insert_pin(class_pin)
+        # finally, delete the command sent to us
+        await ctx.message.delete()
+
+    @BaseService.Listener(Events.on_reaction_add)
+    async def on_reaction_add(self, react: discord.Reaction, user: Union[discord.User, discord.Member]):
+        if react.emoji != PIN_REACTION:
+            return
+        if not (class_pin := await self.pin_repo.get_pin_from_sockbot(react.message)):
+            return
+        if react.count >= MIN_PIN_REACTIONS or Staff.is_staff(user):
+            channel = react.message.channel
+            # make sure our message to-be-pinned still exists
+            if not (to_pin := await fetch_optional_message(channel, class_pin.user_message_id)):
+                await react.message.delete()
+                await self.pin_repo.delete_pin(class_pin)
+                return
+            # check if we have enough space to pin
+            pinned_messages = await channel.pins()
+            # unpin the oldest pin if there is no room
+            if len(pinned_messages) == MAX_PINS_PER_CHANNEL:
+                await pinned_messages[0].unpin(reason='Unpinned to make room.')
+            # pin the message, update the pin in the db, and delete sockbot's pin request embed
+            await to_pin.pin(reason='Pinned by `pin` command.')
```

Comment:
What about specifying the person who started the vote? Also I'm not sure if markdown works in the audit log

Reply:
Good idea, also figured out markdown does not work in the audit log lmao